### ISSUE TITLE
small enhancement possible (replace function ptr with std::function ) 

### ISSUE THREAD

**Issue by nicolas800** [tristanpenman/valijson#30]
Hello

Thank you for your great c++ json schema checker. Please let me suggest one small enhancement 
in the SchemaParser::FunctionPtrs : using function pointer prevent us to use any closure, 

so replacing : 
```
template<typename AdapterType>
struct FunctionPtrs
{
    typedef typename adapters::AdapterTraits<AdapterType>::DocumentType
            DocumentType;

    /// Templated function pointer type for fetching remote documents
    typedef const DocumentType * (*FetchDoc)(const std::string &uri);

    /// Templated function pointer type for freeing fetched documents
    typedef void (*FreeDoc)(const DocumentType *);
};
```

with 
```
#include <functional>
template<typename AdapterType>
struct FunctionObjs
{
    typedef typename adapters::AdapterTraits<AdapterType>::DocumentType
            DocumentType;

    /// Templated function pointer type for fetching remote documents
    typedef const std::function< DocumentType* (const std::string &uri) >  FetchDoc ;

    /// Templated function pointer type for freeing fetched documents
    typedef std::function< void (const DocumentType *)> FreeDoc ;
};
```
would help to use lambda or whatever bind(.. ) structure.

best regards and keep up good job !

Nicolas

**Comment by tristanpenman** [comment]
Thanks for the suggestion, Nicolas. This feature was originally implemented using function pointers so that it would be compatible with non-C++11 compilers. With the recent changes to remove Boost as a hard dependency, it makes sense to change it and embrace the future.

Personally, I won't have time to make this change for a few days, but I'm more than happy to review a and merge a PR if you would like the feature sooner.

**Comment by tristanpenman** [comment]
The only other thing I would consider for this change is whether it is better to keep the current name for the container struct. Preserving the name means that users may not need to change their code at all, and will reduce the number of changes necessary in Valijson itself.

**Comment by nicolas800** [comment]
Indeed !

**Comment by nicolas800** [comment]
Now is we i will try implement it next week and have unit test compile as it.
Nicolas

**Comment by nicolas800** [comment]
Hello

I just pulled the request.

Best regards

Nicolas

**Comment by tristanpenman** [comment]
Thanks for the PR, Nicolas.

Looks nice and clean and has been merged in commit 60eba9d.

**Comment by nicolas800** [comment]
Thanks 
I merged it back on my project.
Nicolas
